Step3: 로또(자동)

mannue opened the pull request:
# 실행방법
### Application 의 main 을 실행시키면 됩니다 😄 

--------------

### class 역할
- Presenter 
  - 입력과 출력을 담당한다.
  - 입력은 로또 구매 금액 , 지난 주 당첨 번호 
  - 출력은 입력 멘트 , 금액에 따른 로또 장수, 구매한 로또 번호 , 지난 주 당첨 번호 멘트, 당첨 통계 정보

- LottoNumber
  - LottoNumber 는 VO 객체이며, 1 ~ 45 까지 정수 값만 가진다.
  - 음수 등 잘못된 값이 입력되면 RuntimeException 예외가 발생한다.
  - LottoNumber 는 비교 할수 있다.

- Lotto
  - 로또는 LottoNumber 숫자 6개 를 가지며, 6개가 아니면 RuntimeException 을 발생한다.
  - 로또는 입력한 LottoNumber 숫자들이 정렬 한 후 저장한다.
  - 로또는 check 메소드는 당첨 번호 Lotto 를 입력으로 받고 입력된 값을 통해서 일치한 갯수를 반환한다.

- Hit
  - Enum 으로써 일치한 갯수 값과 그에 맞는 금액 정보를 가진다.
  - Hit 의 일치한 갯수 값은 0 과 6 사이 값을 가진다.
  - 비교 메소드를 통해서 값을 비교할수 있다.
  - 동일한 당첨금에 갯수를 입력 시 당첨금 * 갯수 에 대한 금액을 반환한다. 
  
- Total
  - Hit[] 입력을 받고 result 메소드를 통해서 EarningsRate 를 반환한다.
  - Total 은 Hit 의 THREE 부터 정보를 저장한다.

- LottoPaper
  - 일급 객체로써 Lotto[] 를 입력 받는다.
  - checkAll 메소드를 통해서 정중한 통계 정보를 반환한다.

- EarningsRate
  - 통계 정보로써 총당첨금 과 사용자 구매 금액 정보를 입력 받는다.
  - rate 정보를 멤버 변수로 가진다.

- LottoVendingMachine
  - 사용자가 로또 구매할 금액을 입력하면 LottoPaper 를 반환한다.
  - 로또 한장의 가격은 1000 원이며 1000원 단위로 Lotto 가 발급된다.

- LottoGenerator
  - create 메소드를 통해서 숫자가 6개인 Lotto 를 생성한다. 


바쁘신 시간 속에서 리뷰 감사합니다.
오늘 하루도 좋은 하루 되세요
감사합니다.

mannue committed:
docs(docs): 로또 구현하기 위한 문서 정리

- 요구 사항 정리
- 프로그래밍 제약 사항
- class 역할

mannue committed:
feat(LottoNumber): LottoNumber 는 정수 값을 가진다.

mannue committed:
refactor(LottoNumber): 테스트 케이스 추가 로 인하여 리팩토링

mannue committed:
feat(LottoNumber): 입력 값이 1보다 작거나 46 이상이면 RuntimeException 이 발생한다

mannue committed:
refactor(LottoNumber): 최소 와 최대 값을 추출

mannue committed:
feat(Hit): Hit 는 일치 갯수 와 당첨금을 가진 enum 이다.

mannue committed:
feat(Hit): Hit 는 일치 갯수 와 당첨금을 가진 enum 이다.

mannue committed:
docs(docs): Lotto 역할 추가

mannue committed:
feat(Lotto): Lotto 는 6개의 숫자를 입력 받아야 된다.

mannue committed:
test(Lotto): 테스트 케이스 추가 로 인하여 리팩토링

mannue committed:
docs(LottoNumber): 비교 기능 추가

mannue committed:
feat(LottoNumber): 서로 비교 할수 있도록 Comparable 상속 받아 구현

mannue committed:
test(LottoNumber): 테스트 케이스 추가

mannue committed:
feat(Lotto): 로또는 입력값을 정렬해서 저장한다.

mannue committed:
refactor(Lotto): 상수 값 추출 과 멤버변수 사용

mannue committed:
test(LottoTest): sortedTest 케이스 추가

mannue committed:
docs(docs): Hit 역할 기능 추가

mannue committed:
feat(Hit): valueOf 메소드 추가

mannue committed:
feat(Lotto): checkTo 메소드에 당첨 Lotto 를 입력하면 일치한 정보를 반환한다.

mannue committed:
style(ALL): 불필요한 import  등 코드 정리

mannue committed:
docs(docs): Statistics 역할 추가

mannue committed:
feat(Hit): 서로 비교 함수 추가

mannue committed:
feat(Statistics): Statistics class 는 Hit.THREE 부터 값을 저장한다.

mannue committed:
docs(Hit): 비교 역할 기능 문서에 추가

mannue committed:
docs(docs): class 이름 변경 과 역할 정리

mannue committed:
feat(Total): Total class 은 Hit THREE 부터 정보를 저장한다.

mannue committed:
docs(docs): 당청금의 갯수만큼 비용을 반환한다.

mannue committed:
feat(Hit): 당청금의 갯수만큼 비용을 반환한다.

mannue committed:
feat(EarningsRate): 총당첨금 과 사용자 구매 금액 정보를 입력 받는다.

mannue committed:
test(EarningsRate): 수익률이 같으면 같은 객체이다.

mannue committed:
test(EarningsRate): 출력 메시지 테스트

mannue committed:
feat(Total): result 메소드를 통해서 EarningsRate 를 반환한다.

mannue committed:
test(Total): Total 출력 테스트

mannue committed:
test(Total): result 테스트 케이스 정리

mannue committed:
feat(LottoPaper): 생성시 Lotto[] 를 입력 받고 입력값이 잘못된 경우 에러를 발생한다.

mannue committed:
feat(LottoPaper): 당첨번호 인 Lotto 를 입력하고 결과로 Total 를 반환한다.

mannue committed:
test(LottoPaper): checkAll 메소드 호출시 인자값은 null 이 될수 없다.

mannue committed:
docs(docs): 로또 자판기 역할 정의

mannue committed:
feat(LottoPaper): 로또 발급 장수를 확인하는 메소드 추가

mannue committed:
feat(LottoVendingMachine): issued 메소드 구현

mannue committed:
docs(docs): LottoGenerator 역할 정의

mannue committed:
feat(LottoGenerator): 자동 발급과 수동발급 메소드 구현

mannue committed:
refactor(LottoPaper): 불필요한 메소드 삭제

mannue committed:
fix(Total): set 은 순서를 보장하지 않아서 출력이 바뀌는 문제 발생

key 를 정렬하여 가져오도록 수정

mannue committed:
refactor(LottoVendingMachine): Lotto 생성하는 부분을 LottoGenerator 로 역할 이전

mannue committed:
test(LottoVendingMachine): 테스트 안에 중복 제거 와 잘못된 입력값 테스트

mannue committed:
feat(LottoPaper): 로또 구매한 장수를 알수 있는 메소드 추가

mannue committed:
feat(Lotto, LottoNumber, LottoPaper): toString 메서드 추가

mannue committed:
feat(Presenter): 화면상 입력 과 출력을 담당하는 클래스

mannue committed:
feat(Lotto): 셍성자 추가

mannue committed:
feat(Presenter): 출력 메소드 수정

mannue committed:
fix(Total): hit 값이 없어도 다 출력해야 된다.

mannue committed:
feat(LottoService): service class 구현

mannue committed:
feat(Application): 시작프로그램 구현

mannue committed:
refactor(LottoService): Lotto 생성시 LottoGenerator 를 사용하도록 수정

mannue committed:
feat(Hit): 당첨리스트를 반환하는 함수 추가

mannue committed:
refactor(Total): 불필요한 초기화 부분 제거

mannue committed:
fix(LottoGenerator): 수동 생성시 중복 제거 추가

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/EarningsRate.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.stream.IntStream;
+import java.util.stream.LongStream;
+
+public class EarningsRate {
```
극단적인 객체 분리 👍

mannue replied:
감사합니다~

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/EarningsRate.java:
```diff
@@ -0,0 +1,47 @@
+        this.rate =  Math.floor((totalPrizeAmount / userBuyAmount) * 100) / 100.0;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        EarningsRate that = (EarningsRate) o;
+        return Double.compare(that.rate, rate) == 0;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(rate);
+    }
+
+    @Override
+    public String toString() {
+       return String.format("총 수익률은 %.2f입니다.%s",this.rate, messageType());
+    }
```
비지니스와 상관 없는 메소드들은 클래스 최하단에 위치시키는것은 어떨까요?

mannue replied:
messageType() 도 비지니스 와 상관없어서 최하단에 위치했는데요 
혹시 toString -> messageType -> equals -> hasCode 이 순위로 되어야 되는걸까요?

seok-2-o replied:
지금은 콘솔 뷰를 사용하고 있기때문에 messageType() 는 어플리케이션에서 상관이 있지 않을까요 ?
추가로 콘솔 뷰가 아닌 다른 뷰(웹/앱) 을 사용한다면 toString 과 messageType 는 이용하기 어려울 것 같아요.

mannue replied:
우선 메소드 이름도 변경하였습니다. 
messageType 메소드가 수익률 관련해서 추가 설명하는 메세지 여서 좀더 그에 맞게 이름을 아래와 같이 수정하였습니다.
explanation 😄

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/Hit.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.Arrays;
+
+public enum Hit {
+    ZERO(0, 0),
+    ONE(1, 0),
+    TWO(2, 0),
```
```suggestion
    MISS(3, 0),
```
구지 세 상태를 분리할 필요가 있을까요?

mannue replied:
제가 해당 enum 을 만들때 인터넷 로또를 생각하면서 만들었는데요
동X복권 사이트 에서 사용자가 로또 확인하면 안맞거나 , 한개 , 두개 맞아도 보여주자나요 
그런 의미에서 만들었습니다. 그래서 세 상태를 놔뒀습니다~

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/Hit.java:
```diff
@@ -0,0 +1,46 @@
+    THREE(3, 5000),
+    FOUR(4, 50000),
+    FIVE(5, 1500000),
+    ALL(6, 2000000000);
```
```suggestion
    ALL(6, 2_000_000_000);
```
위와 같이 가독성을 향상시킬 수 있어요.

mannue replied:
오 ~ 그럼 단위를 좀 더 쉽게 이해 되겠네요 
감사합니다.
혹시 여기서 한가지 궁금한게 있는데요 한국은 3개 단위로 끊는데요 
외국 같은경우에는 그렇게 하지 않자나요 이런 경우에는 어떻게 대처해야 되는지 의견주시면 감사하겠습니다~ 👍

seok-2-o replied:
외국에서도 금액은 3자리로 끊어서 쓰지 않나요 ㅎ `-` 가 세자리마다 써야 정상적으로 동작하는 것은 아니니
함께 일하는 사람들과 약속된 표현으로 사용하면 좋을것 같아요.

mannue replied:
네 의견 감사합니다 숫자는 끊어서 쓰도록 해야겠네요 ^^

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/Hit.java:
```diff
@@ -0,0 +1,46 @@
+        return target.isHigh(this.count);
+    }
+
+    int cost(final int count) {
```
`Hit.All.cost(2)` 하면 어떻게 될까요?

mannue replied:
cost 는 당첨갯수에 따른 당첨금을 주도록 하는 메소드로 만들었는데요
그래서 All.cost(2) 하면 1등 2장 이라는 개념이 됩니다.

seok-2-o replied:
int 의 크기는 20억 * 2를 담을 수 있을까요?

mannue replied:
long 형으로 변경하였습니다.
조언 감사합니다 👍

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/Lotto.java:
```diff
@@ -0,0 +1,56 @@
+    private static final int VALID_SIZE = 6;
+    private List<LottoNumber> lottoNumbers;
+
+    public Lotto(LottoNumber[] lottoNumbers) {
```
중복된 번호로 로또가 생성 가능할 것 같아요.

seok-2-o replied:
중복을 허용하지 않는 자료구조를 다룬다면 더 좋은 자료구조를 고민해보는것도 좋을 것 같아요.

mannue replied:
생성자에 리스트로 받도록 수정하였으며, 중복체크 후 유효성 검사하도록 하였습니다.

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/LottoGenerator.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    private void initLottoNumbers() {
+        final int START_NUMBER = 1;
+        final int END_NUMBER = 45;
```
대문자로 선언하신 이유가 있으신가요?

mannue replied:
상수라서 대문자로 했지만 좀더 변수 이름을 쉽게 알아보기 위해서 minNumber, maxNumber 로 하였습니다.

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/LottoGenerator.java:
```diff
@@ -0,0 +1,35 @@
+        return lottoArray;
+    }
+
+    public Lotto manual(final int[] numbers) {
```
현재 요구사항이 아닌 것을 미리 구현하신 이유가 있으신가요?

본과정은 현업에서 요구사항이 추가되는 부분을 감안하여 설계되어 있습니다.

mannue replied:
이부분은 사용자가 저번주 로또 번호를 입력할때 수동으로 입력하기 때문에 manual 이라는 메소드를 만들게 되었습니다.

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/LottoGenerator.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public Lotto manual(final int[] numbers) {
+        return new Lotto(Arrays.stream(numbers).mapToObj(LottoNumber::new).distinct().toArray(LottoNumber[]::new));
```
배열을 사용하시는 이유가 있으실까요?

mannue replied:
사용자가 입력한 수동 번호를 그대로 사용하기 위해서 배열을 사용했는데요 
음... 이부분도 List 로 바꿔야 겠네요 
그리고 사용자 입력시 배열을 리스트로 변환하고 변경하도록 하겠습니다.
왠만하면 배열은 자체하도록 하겠습니다 😄 
감사합니다~

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/LottoNumber.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.Objects;
+
+public class LottoNumber implements Comparable<LottoNumber> {
+    private final int number;
+    private static final int MIN = 1;
+    private static final int MAX = 45;
```
클린코드에서 나오는 클래스 선언의 표준 관례는 아래와 같습니다.

정적 공개 상수 (public static final)
정적 비공개 상수 (private static final)
비공개 인스턴스 변수(private) - 공개 인스턴스 변수는 사용하지 않는다.
공개 함수(public ... method())
비공개 함수(private ... method())

mannue replied:
표준관례를 지키도록 하겠습니다 ~ 감사합니다 👍 
감사합니다 ^^

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/LottoPaper.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+public class LottoPaper {
+    private final List<Lotto> userLottoList;
```
변수명의 자료구조를 명시하면 나중에 자료구조가 변경되어야 하는 경우 수정할 부분이 많아 질 수 있어요.

mannue replied:
Lotto 자료형이 바뀌면 변수이름도 변경되겠네요 
무의식처럼 이름을 쉽게 만들었는데 좀 더 신경써야 될것 같습니다 감사합니다 😄 

구매리스트 뜻으로 purchaseList 로 변경하려고 합니다 ^^
다시 한번 좋은 조언 감사합니다~

seok-2-o replied:
제가  변수명에 List 가 들아간 부분에 대해서 말씀드렸습니다.
그럴일은 없지만 자료구조가 Set<Lotto> 로 변경된다면 변수명도 변경해줘야 할 것 같아요.

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/Total.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.Map;
+import java.util.Objects;
+
+public class Total {
```
Total이라는 클래스 이름이 무엇을 하는지 나타내기에는 조금 아쉬운 것 같아요.

mannue replied:
Total 이름으로 생각하기에는 @seok-2-o 님 말씀처럼 애해 한것 같습니다.
그래서 Result 로 할려고 하니 차라리 로또 결과 정보니 LottoResult 로 변경하였습니다.
의견 감사합니다 👍

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/Total.java:
```diff
@@ -0,0 +1,57 @@
+public class Total {
+    private final Map<Hit, Integer> totalHitMap = new HashMap<>();
+
+    public Total(Hit... hits) {
```
전반적으로 가변인자를 사용하신 이유가 궁금하네요.

seok-2-o replied:
[[Effective java] item 28. 배열보다는 리스트를 사용하라](https://velog.io/@new_wisdom/Effective-java-item-28.-%EB%B0%B0%EC%97%B4%EB%B3%B4%EB%8B%A4%EB%8A%94-%EB%A6%AC%EC%8A%A4%ED%8A%B8%EB%A5%BC-%EC%82%AC%EC%9A%A9%ED%95%98%EB%9D%BC)

mannue replied:
제가 가변인자를 사용한것은 테스트 시에 항상 배열은 만드는게 쉽지 않아서 사용하게 되었네요 
https://jackjeong.tistory.com/125
글도 읽어보니 가변인자는 성능에도 문제가 있어 보입니다. 좋은 조언 감사합니다 👍 

배열보다는 리스트를 사용하라 이부분에서 생성자시에는 배열로 받고 내부적으로는 리스트를 사용하는건 
이것또한 사용안하는게 좋을까요?

seok-2-o replied:
배열을 무조건 사용하면 안되는것은 아니지만 배열보다 리스트로 쓰는것이 장점이 많기때문에 저는 리스트를 추천드립니다.

mannue replied:
네 알겠습니다. 배열보다는 리스트를 사용하도록 하겠습니다. 👍

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/Hit.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public static Hit[] winningList() {
+        return Arrays.stream(Hit.values()).filter(Hit.TWO::isLow).toArray(Hit[]::new);
```
```suggestion
        return Arrays.stream(Hit.values())
        .filter(Hit.TWO::isLow)
        .toArray(Hit[]::new);
```
스트림은 행위 단위로 내려쓰는 것이 가독성과 디버깅 차원에서 좋습니다.

mannue replied:
동료들이 보기에도 내리는게 좋을것 같네요 👍 
내려 쓰도록 연습해봐야겠네요 
감사합니다 😄

mannue replied:
동료들이 보기에도 내리는게 좋을것 같네요 👍
내려 쓰도록 연습해봐야겠네요
감사합니다

seok-2-o reviewed (changes requested):
로또 미션 정말 잘 진행해주셨네요 👍 
극단적인 객체분리가 정말 인상적이네요.

몇가지 고민거리 남겨드렸으니 확인해보시면 좋을 것 같아요.

우테캠프로와 함께하는 즐거운 주말 되시길바라요 🙏

mannue committed:
refactor(LottoNumber): 표준관례 로 순서를 유지하자!

정적 공개 상수 (public static final)
정적 비공개 상수 (private static final)
비공개 인스턴스
변수(private) - 공개 인스턴스 변수는 사용하지 않는다.
공개 함수(public ... method())
비공개
함수(private ... method())

mannue committed:
refactor(LottoPaper, Total): 가변인자 사용을 자제하자 !!

가변인자는 호출시 배열을 생성하니 성능상 문제가 될수 있다.

mannue committed:
refactor(Hit): 숫자 가독성을 위해서 _ 를 사용하자

mannue committed:
refactor(Hit, LottoGenerator): stream 가독성을 위해서 줄내림

mannue committed:
refactor(LottoResult): Total -> LottoResult 로 class 이름 변경

total 보다 LottoResult 로 좀더 명확하게 하는게 좋은것 같습니다.

mannue commented:
조언 주신 부분 과 언급하신 부분에 대해서 comment 작성 하였습니다.
바쁘신 가운데 꼼꼼하게 리뷰 해주셔서 감사합니다 😄 👍

seok-2-o reviewed (changes requested):
-

seok-2-o reviewed (changes requested):
피드백 반영 잘해주셨네요 👍 
다만 확인 되지 않은 피드백이 있는것 같아서 다시 확인 요청드러요.
![image](https://user-images.githubusercontent.com/34270397/168420344-03912eaa-d805-4449-984f-2c95798af7dc.png)

질문하신 부분에는 답변 달아놓았습니다.

mannue committed:
refactor(Hit): 타입 변경 과 메소드 이름 변경

mannue committed:
feat(LottoMoney): 로또 구입 금액 정보를 가진 클래스 생성

mannue committed:
feat(LottoNumbers): 1 ~ 45 까지 로또 숫자들을 가진 클래스

mannue committed:
refactor(LottoResult): 수익률 계산을 LottoMoney class 에 이관 , 총 상금 계산 결과 타입 변경

mannue committed:
refactor(LottoNumbers): 생성자 추가

mannue committed:
refactor(EarningsRate): 수익률 정보만 갖도록 수정

mannue committed:
refactor(LottoPaper): 생성자 인자 타입 수정

mannue committed:
refactor(LottoGenerator): LottoNumbers 를통해서 로또번호를 추출하도록 수정

mannue committed:
refactor(Presenter): 사용자가 로또 구매금액 입력시 LottoMoney 를 반환하도록 수정

mannue committed:
refactor(LottoVendingMachine): LottoMoney를 이용하여 금액만큼 자동으로 뽑도록 수정

mannue committed:
refactor(HitTest): 테스트 수정

mannue committed:
refactor(LottoResultTest): LottoMoney 를 사용하도록 수정

mannue committed:
refactor(LottoService): LottoMoney 를 사용하도록 수정

mannue committed:
style(Hit , LottoMoney): 코드 스타일 정리

mannue committed:
refactor(EarningsRate): 메소드 이름 수정

mannue committed:
refactor(All): Lotto 생성자 타입 변경

mannue committed:
refactor(All): 불필요한 코드 정리

mannue committed:
docs(docs): 문서정리

mannue commented:
@seok-2-o  님 안녕하세요
정성스럽게 리뷰 해주셔서 감사합니다.

###  수정 내역 
-  생성자 안에서 배열보다 리스트로 변경
- LottoMoney  원시값 보장
  - 사용자가 로또 구매할 금액을 가지고 있다.
  - 총 당청금액을 입력하면 총 수익률을 반환한다.
- 로또 추출은 LottoNumbers 에 이관
- EarningsRate 는 수익률 정보만 가지도록 수정
- LottoGenerator 는 수동 과 자동 메소드만 제공하도록 수정


죄송합니다. 하다보니 많은 부분이 고쳐 졌네요
바쁘신 시간 중에 감사드리며, PR 합니다.
주말 잘 보내세요~ 👍

mannue reviewed (commented):
manual 메소드를 만들게 된 배경에 대해서 설명 추가하였습니다~
리뷰 감사 합니다~

mannue committed:
fix(Lotto): 중복 제거후 유효성 검사

mannue committed:
refactor(LottoNumbers): 변수 이름 변경

mannue reviewed (commented):
바쁜 시간속에서 좋은 조언 감사합니다. 👍

mannue reviewed (commented):
사용자가 입력한 수동 번호를 그대로 사용하기 위해서 배열을 사용했는데요
음... 이부분도 List 로 바꿔야 겠네요
그리고 사용자 입력시 배열을 리스트로 변환하고 변경하도록 하겠습니다.
왠만하면 배열은 자체하도록 하겠습니다 😄
감사합니다~

mannue committed:
refactor(LottoGenerator): manual 시 List 로 인작값 변경

mannue reviewed (commented):
좋은 리뷰 와 꼼꼼하게 봐주셔서 감사합니다 👍

mannue committed:
style(Lotto): 메소드 이름 변경

mannue committed:
refactor(LottoGenerator): Lotto를 생성하는 역할은 LottoGenerator만 한다.

mannue committed:
refactor(LottoNumbers): 추출된 로또 번호들을 수정이 불가능 해야 된다.

mannue committed:
refactor(LottoNumbers): spy 를 통해서 테스트용으로 사용하는 생성자 제거

mannue committed:
test(LottoGeneratorTest): spy의 인자값 타입을 좀더 명시적으로 정하고 행위 검증 추가

mannue committed:
style(LottoGenerator): 불필요한 코드 제거

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/Lotto.java:
```diff
@@ -0,0 +1,56 @@
+
+public class Lotto {
+    private static final int VALID_SIZE = 6;
+    private final List<LottoNumber> lottoNumbers = new ArrayList<>();
```
중복되지 않는 숫자르 다루는 더 좋은 자료형은 존재하지 않을까요?

mannue replied:
private final Set<LottoNumber> lottoNumbers = new TreeSet<>();

    public Lotto(final List<LottoNumber> lottoNumbers) {
        this(new HashSet<>(lottoNumbers));
    }

    public Lotto(final Set<LottoNumber> lottoNumbers) {
        validationSize(lottoNumbers);
        this.lottoNumbers.addAll(lottoNumbers);
    }

조언 감사합니다 Set 으로 생각하다가 Collection.sortd 가 List 여서 이거 사용해야 되는줄 알고 List 로 했네요

TreeSet 으로 사용하면 정렬도 되고 중복도 제거 되어서 TreeSet 으로 변경하였습니다~

seok-2-o replied:
정렬은 출력할때만 필요하니 도메인과  DTO 를 분리하는것도 한 방법일것 같아요.

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/Lotto.java:
```diff
@@ -0,0 +1,56 @@
+    private int checkBy(List<LottoNumber> userLottoNumbers) {
+        return userLottoNumbers.stream()
+                .filter(this.lottoNumbers::contains)
+                .toArray().length;
```
```suggestion
.count();
```

mannue replied:
return Long.valueOf(userLottoNumbers.stream()
                .filter(this.lottoNumbers::contains)
                .count())
                .intValue();

index 가 int 라서 count 가 long 형이라서 int 로 변경해줘야 되서 위 코드로 변경하였습니다.

toArray.length 보다 좋은것 같습니다.
조언 감사합니다 👍

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/LottoNumbers.java:
```diff
@@ -0,0 +1,53 @@
+
+    public final List<LottoNumber> extract(Consumer<List<LottoNumber>> consumer) {
+        consumer.accept(numbers);
+        return Collections.unmodifiableList(numbers.subList(0, 6));
```
매직넘버는 상수로 분리하는게 좋을것 같아요.
로또의 갯수가 6에서 변경될 일은 크게 없긴 하지만 비지니스적인 요구사항이므로 따로  관리하는게 좋지 않을까요

그 외에도 상수들을 찾아 분리해면 좋을 것 같아요.

mannue replied:
네 상수들은 따로 분리하도록 하겠습니다.
조언 감사합니다 👍

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/LottoNumbers.java:
```diff
@@ -0,0 +1,53 @@
+public class LottoNumbers {
+    private final List<LottoNumber> numbers = new ArrayList<>();
+
+    public LottoNumbers() {
+        final int minNumber = 1;
+        final int maxNumber = 45;
+        for (int i = minNumber; i <= maxNumber; i++) {
+            numbers.add(new LottoNumber(i));
+        }
+    }
```
매번 1~45개의 로또 번호를 미리 생성해놓을 필요가 있을까요?

mannue replied:
우선 저는 로또 번호를 1 부터 45까지 존재해야 된다고 미리 생성했습니다.
저는 이 클래스를 로또 뽑기 위한 번호들로 블랙박스 처럼 외부에서 변경 못하도록 하고 싶어서 만들었습니다 ^^;;

seok-2-o replied:
LottoNumbers를 한번만 생성하는거였군요 제가 착각했습니다 ~

mannue replied:
아닙니다 제가 좀 더 잘 이해할수 있도록 코드를 작성 했어야 되는데 아직 많이 부족하네요 ^^;
리뷰 감사합니다 ^^

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/LottoNumbers.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.Objects;
+import java.util.function.Consumer;
+
+public class LottoNumbers {
```
해당 클래스는 객체인가요?

mannue replied:
저는 해당 클래스를 블랙박스처럼 외부에서는 로또 뽑기 위한 번호 리스트 들을 절대 수정 못하게 하기 위해서 해당 클래스를 만들었습니다.
음.. 마치 한번 만들면 누구도 이안의 정보는 변경하지 못한다 라는 느낌으로요 ㅎㅎ

@seok-2-o  님 더 좋은 방법이나 의견 주시면 감사하겠습니다 😄

seok-2-o replied:
그렇다면 정적 메소드로 만드는게 더 좋지 않을까요?

mannue replied:
로또 기계들이 여러개라고 생각하면  LottoNumbers 가 오염되면 모든 기계에 영향을 주기 때문에 현재 상태로 있는게 낫지 않을까 
생각해 봅니다 
로또 숫자들이 있는 바구니가 한개 있는데 여러군데에서 사용하면 문제가 될것 같다는 생각을 해봅니다 😄

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/LottoGenerator.java:
```diff
@@ -0,0 +1,24 @@
+        this.numbers = numbers;
+    }
+
+    public Lotto auto() {
+        return new Lotto(numbers.extract(Collections::shuffle));
+    }
+
+    public Lotto manual(final List<LottoNumber> numbers) {
+        return new Lotto(numbers);
+    }
```
numbers 의 기능은 해당 객체로 이동시키는것에대해서 어떻게 생각하시나요?

mannue replied:
nubers는 로또 번호들중에서 뽑는 방법을 통해서 번호만 추출하도록 하기 위해서 만들 class 입니다.
반면에 LottoGenerator는 로또 번호들을 통해서 로또를 생성하기 위한 역할로 만들었습니다. 자동은 시스템상 자동으로 로또를 만들기 위한 메소드 이고 manual 은 사용자가 입력한 번호들은 만드는 메소드롤 만들었습니다.

지난주 로또 번호를 사용자가 입력하자나요 거기서 이부분은 수동이라고 저는 생각했습니다.

짧게 저의 생각을 적어보았습니다. 글쓰는 재주가 없어서 죄송합니다 ^^;

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/LottoPaper.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.stream.Collectors;
+
+public class LottoPaper {
+    private final List<Lotto> purchaseList;
```
List<Lotto> purchase`List`; 
변수 이름에 자료구조형을 명시하는것은 안티패턴입니다.

mannue replied:
네 알겠습니다. 😄 
list 이름을 제거 하였습니다~ 👍

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/LottoResult.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    private void increase(final Hit hit) {
+        if (Hit.TWO.isLow(hit)) {
```
`hit.isWin()`
은 어떤가요?

mannue replied:
가독성 면에서 isWin 이 좋을듯 합니다.  
조언 감사합니다 👍

seok-2-o commented on src/main/java/camp/nextstep/edu/step3/LottoResult.java:
```diff
@@ -0,0 +1,61 @@
+        return message.toString();
+    }
+
+    private String printFormat(final Hit hit) {
+        return String.format(hit + "- %d개\n", resultMap.getOrDefault(hit, 0));
+    }
```
도메인에서 출력을 담당하면 
CosoleView가 아닌 다른 View 를 이용할 경우 어떻게 될까요?

mannue replied:
우선 제가 이렇게 만든 이유가 도메인 에서 정보를 추출 하지 말라는 생각으로 만들었는데요
View 는 출력만 담당하자 라는 생각했습니다.
만약 View 가 Hit 갯수당 출력을 담당해야 된다면 LottoResult 에 getHitCount() 를 만들어 줘야 되는데 ... 출력 을 위해서 해당 메소드를 만들어야 되는지 의문이 드네요 ^^;

이부분에 대해서 저의 생각을 적어보았습니다.
의견이나 조언 부탁드립니다 ^^

seok-2-o replied:
getter를 완전히 피할수는 없습니다. 출력을 위해서는 당연히 getter를 만들어줘야 한다고 생각합니다.

웹이나 앱에서 해당 어플레키에션이 동작할 수 있을까요?

mannue replied:
음 도메인의 출력 format 을 toString 에서 해결하면 웹이나 앱에서도  toString 으로 안에 내용은 출력 가능 하지 않을까요?
그래서 String 으로 하도록 했는데요. 저는 가능 할것 이라고 생각하고 있는데요 
혹시 좋은 방법이나 경험 알려주시면 감사하겠습니다 👍 😃

seok-2-o reviewed (changes requested):
거의 새로 만들다시피 하셨네요 👍 
쉽지 않으신 결정이셨을텐데 고생많으셨습니다.

몇가지 고민거리 남겨드렸으니 확인 부탁드릴께요.

mannue committed:
feat(Lotto): 자료형 변경

mannue committed:
refactor(Lotto): stream toArray.length  보다 count 를 사용하자!

mannue committed:
style(LottoPaPer): 멤버변수 이름 변경

변수 이름에 자료구조형을 명시하는것은 안티패턴입니다.

mannue committed:
style(Hit): Hit isLow 를 isHigh 로 수정

mannue committed:
style(Hit): Hit isLow 를 isWin 로 수정

mannue committed:
style(LottoNumbers): 상수를 분리 추출

mannue reviewed (commented):
@seok-2-o  님의 꼼꼼한 리뷰에 감사드립니다 👍 
저의 생각과 조언 주신 부분은 수정하였습니다.
주말 잘 보내세요~

mannue committed:
style(Lotto, LottoResult): 멤버변수 이름변경

mannue committed:
style(LottoResult): 멤버변수 이름변경

seok-2-o reviewed (approved):
빠르게 피드백 주셨네요 👍 
질문하신 부분에 답변 및 코멘트 남겨놓았습닏.

이어서 다음 미션 진행해주세요.
